optuna/multi_objective/trial.py

```diff
@@ -0,0 +1,376 @@
+from datetime import datetime
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Sequence
+from typing import Union
+
+from optuna._experimental import experimental
+from optuna.distributions import BaseDistribution
+from optuna import multi_objective
+from optuna.study import StudyDirection
+from optuna.trial import FrozenTrial
+from optuna.trial import Trial
+from optuna.trial import TrialState
+
+CategoricalChoiceType = Union[None, bool, int, float, str]
+
+
+@experimental("1.4.0")
+class MultiObjectiveTrial(object):
+    """A trial is a process of evaluating an objective function.
+
+    This object is passed to an objective function and provides interfaces to get parameter
+    suggestion, manage the trial's state, and set/get user-defined attributes of the trial.
+
+    Note that the direct use of this constructor is not recommended.
+    This object is seamlessly instantiated and passed to the objective function behind
+    the :func:`optuna.multi_objective.study.MultiObjectiveStudy.optimize()` method;
+    hence library users do not care about instantiation of this object.
+
+    Args:
+        trial:
+            A :class:`~optuna.trial.Trial` object.
+    """
+
+    def __init__(self, trial: Trial):
+        self._trial = trial
+        self._n_objectives = multi_objective.study.MultiObjectiveStudy(trial.study).n_objectives
```

Comment:
Not a strong opinion but the trial could also hold its number of objectives in its system attributes (e.g. set in the `mo_objective` wrapper function). Then, we don't have to construct a `MultiObjectiveStudy` each time we get here.

Reply:
Thank you for your opinion. I agree that there is definitely room of improvement. Personally, it feels sufficient to store the number of objectives to the system attribute of the study (instead of storing it to the system attribute of each trial).
For now, I left a TODO comment here (https://github.com/optuna/optuna/pull/1054/commits/03e486b5ec72d0f14b40d31027ccaf670b7a8010). Let's discuss this topic separately from this PR.